midend/predication.cpp

```diff
@@ -15,21 +15,26 @@ limitations under the License.
 */
 #include "predication.h"
 #include "frontends/p4/cloner.h"
-
-
 namespace P4 {
-/// convert an expression into a string that uniqely identifies the value referenced
-/// return null cstring if not a reference to a constant thing.
-static cstring expr_name(const IR::Expression *exp) {
+/// convert an expression into a string that uniqely identifies the lvalue referenced
+/// return null cstring if not a reference to a lvalue.
+static cstring lvalue_name(const IR::Expression *exp) {
     if (auto p = exp->to<IR::PathExpression>())
         return p->path->name;
     if (auto m = exp->to<IR::Member>()) {
-        if (auto base = expr_name(m->expr))
+        if (auto base = lvalue_name(m->expr))
             return base + "." + m->member;
     } else if (auto a = exp->to<IR::ArrayIndex>()) {
         if (auto k = a->right->to<IR::Constant>())
-            if (auto base = expr_name(a->left))
-                return base + "." + std::to_string(k->asInt()); }
+            if (auto base = lvalue_name(a->left))
+                return base + "." + std::to_string(k->asInt());
+    } else if (auto s = exp->to<IR::Slice>()) {
+        if (auto base = lvalue_name(s->e0))
+            if (auto h = s->e1->to<IR::Constant>())
+                if (auto l = s->e2->to<IR::Constant>())
+                    return base + "." +
+                        std::to_string(h->asInt()) + ":" + std::to_string(l->asInt());
+    }
```

Comment:
If "" should never happen then you should make this a BUG_CHECK.

Reply:
lvalue_name should return "" cstring for every expression that is not lvalue.